### PR TITLE
[feat] refactor data storage to allow saving on a per-window basis

### PR DESCRIPTION
**Pull request by shouldsee**
## Description

- changed eager mode logic 
- changed data saving and loading logic
    - old logic in `SimpleJsonStorage`
    - per-window logic in `SimpleWindowJsonStorage`
    - using hard-coded MRO to switch between the logic for now.
- added a utility script for testing
- added auto directory creation
- added DEBUG utility printing

## Motivation and Context

#861 . changing towards a per-window caching scheme

## How Has This Been Tested?

- `rm -rf ./test_data/*`
- start server `python server.py -env_path test_data`
- `python test_simple.py` create plots an save
- closed server 
- reopen server
- check `env:test` is correctly loaded

## Screenshots (if appropriate):

```
tree test_data/
test_data/
├── test
│   ├── jsons
│   │   ├── 0.json
│   │   ├── 1.json
│   │   ├── 2.json
│   │   ├── 3.json
│   │   └── 4.json
│   └── reload.json
└── view

```

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
- [ ] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: js/TextPane.js [hunk 1]
```diff
@@ -20,7 +20,7 @@ class TextPane extends React.Component {
     switch (e.type) {
       case 'keydown':
       case 'keypress':
-        e.preventDefault();
+//        e.preventDefault();  /// this would prevent entering text into input box
```
Comment (JackUrb) [inline js/TextPane.js:23]: IIRC this was trying to prevent text panes from submitting on enter. Plus, the goal is to send keypress events to the backend, not necessarily to have them render up-front here.

File: py/visdom/Makefile [hunk 2]
```diff
@@ -0,0 +1,21 @@
+clean:
```
Comment (JackUrb) [inline py/visdom/Makefile:1]: I'm somewhat against adding _yet another_ scripting flow to the visdom codebase. This should either be part of the `package.json` scripts, or we should be having a discussion about the alternate we should be using instead.

File: py/visdom/__init__.py [hunk 3]
```diff
@@ -576,12 +577,12 @@ def on_message(ws, message):
                     try:
                         handler(message)
                     except Exception as e:
+                        import traceback
+                        tbmsg = traceback.format_exc()
                         logger.warn(
-                            'Visdom failed to handle a handler for {}: {}'
-                            ''.format(message, e)
+                            'Visdom failed to handle a handler for {}: {}.'
+                            '\n traceback:{}'.format(message, e, tbmsg)
```
Comment (JackUrb) [inline py/visdom/__init__.py:584]: This is a nice-to-have addition

**Review by JackUrb — COMMENTED** [review]
Hi @shouldsee. Thanks for putting up a PR. I'll admit, this one is pretty difficult to review for a number of reasons.
- Class naming here is somewhat obtuse. `B`, `NIF`, etc
- Without explanations for these classes both inline and in some kind of overall explanation, it's hard to follow your intent and methodology.
- Overall, these additions should exist outside the `server.py` file, which is already bloated (and part of why we're trying to refactor).
- Addition of `Makefile` is complicating the devX here.
- Last, there are some comment-ed out changes (like in `demo.py`), which we really shouldn't be maintaining in the codebase, alongside conditionals such as `if 1:`

I'm open to the feature change you're proposing here, but I'm definitely not ready to fully review, let alone accept, in the current state. Would love to see it all in a more final state.

File: js/TextPane.js (hunk 1 above)
Comment (shouldsee) [reply js/TextPane.js:23]: Ah I see. I was putting forms into visdom window, and .preventDefault would make the form inputting fairly laggy, hence commented out.

Comment (JackUrb) [reply js/TextPane.js:23]: Perhaps there's value in a new `FormPane` component, likely a parent class for the `PropertiesPane`. This would then be able to accomplish what you're attempting in #868, but in a cleaner way

**Comment by tonypzy** [comment]
The server has since been refactored into a module structure (py/visdom/server/), making this PR's diff no longer applicable. The per-window storage concept has good sides but it needs a fresh implementation against the current architecture. Closing after 2+ years of inactivity.
